rust/kernel/soc/apple/rtkit.rs

```diff
@@ -0,0 +1,277 @@
+        Err(EINVAL)
+    }
+}
+
+/// Represents `struct apple_rtkit *`.
+///
+/// # Invariants
+///
+/// The rtk pointer is valid.
+/// The data pointer is a valid pointer from T::Data::into_foreign().
+pub struct RtKit<T: Operations> {
+    rtk: *mut bindings::apple_rtkit,
+    data: *mut core::ffi::c_void,
+    _p: PhantomData<T>,
+}
+
+unsafe extern "C" fn crashed_callback<T: Operations>(cookie: *mut core::ffi::c_void) {
+    T::crashed(unsafe { T::Data::borrow(cookie) });
+}
+
+unsafe extern "C" fn recv_message_callback<T: Operations>(
+    cookie: *mut core::ffi::c_void,
+    endpoint: u8,
+    message: u64,
+) {
+    T::recv_message(unsafe { T::Data::borrow(cookie) }, endpoint, message);
+}
+
+unsafe extern "C" fn recv_message_early_callback<T: Operations>(
+    cookie: *mut core::ffi::c_void,
+    endpoint: u8,
+    message: u64,
+) -> bool {
+    T::recv_message_early(unsafe { T::Data::borrow(cookie) }, endpoint, message)
+}
+
+unsafe extern "C" fn shmem_setup_callback<T: Operations>(
+    cookie: *mut core::ffi::c_void,
+    bfr: *mut bindings::apple_rtkit_shmem,
+) -> core::ffi::c_int {
+    // SAFETY: `bfr` is a valid buffer
+    let bfr_mut = unsafe { &mut *bfr };
+
+    let buf = if bfr_mut.iova != 0 {
+        bfr_mut.is_mapped = true;
+        T::shmem_map(
+            // SAFETY: `cookie` came from a previous call to `into_foreign`.
+            unsafe { T::Data::borrow(cookie) },
+            bfr_mut.iova as usize,
+            bfr_mut.size,
+        )
+    } else {
+        bfr_mut.is_mapped = false;
+        // SAFETY: `cookie` came from a previous call to `into_foreign`.
+        T::shmem_alloc(unsafe { T::Data::borrow(cookie) }, bfr_mut.size)
+    };
+
+    let mut buf = match buf {
+        Err(e) => {
+            return e.to_kernel_errno();
+        }
+        Ok(buf) => buf,
+    };
+
+    let iova = match buf.iova() {
+        Err(e) => {
+            return e.to_kernel_errno();
+        }
+        Ok(iova) => iova,
+    };
+
+    let slice = match buf.buf() {
+        Err(e) => {
+            return e.to_kernel_errno();
+        }
+        Ok(slice) => slice,
+    };
+
+    if slice.len() < bfr_mut.size {
+        return ENOMEM.to_kernel_errno();
+    }
+
+    bfr_mut.iova = iova as u64;
+    bfr_mut.buffer = slice.as_mut_ptr() as *mut _;
+
+    // Now box the returned buffer type and stash it in the private pointer of the
+    // `apple_rtkit_shmem` struct for safekeeping.
+    match Box::try_new(buf) {
+        Err(e) => Error::from(e).to_kernel_errno(),
+        Ok(boxed) => {
+            bfr_mut.private = Box::into_raw(boxed) as *mut _;
+            0
+        }
+    }
+}
+
+unsafe extern "C" fn shmem_destroy_callback<T: Operations>(
+    _cookie: *mut core::ffi::c_void,
+    bfr: *mut bindings::apple_rtkit_shmem,
+) {
+    let bfr_mut = unsafe { &mut *bfr };
+    // SAFETY: Per shmem_setup_callback, this has to be a pointer to a Buffer if it is set.
+    if !bfr_mut.private.is_null() {
+        unsafe {
```

Comment:
This `SAFETY` comment should be moved near the `unsafe`, also Markdown.

```suggestion
    if !bfr_mut.private.is_null() {
        // SAFETY: Per `shmem_setup_callback`, this has to be a pointer to a `Buffer` if it is set.
        unsafe {
```